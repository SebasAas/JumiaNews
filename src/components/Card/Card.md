## Importing Component

```js static
import Card from '<RootFolder>';
```

## Basic Card:

```html
<Card />
```

<hr/>

## Card with loading:

### Expected
- String or Object

### Default
- false

```html
<Card loading='true' />
```

<hr/>

## Card with custom size Loader:

### Expected

 - <b>big</b>: String
    - Loader size height 370px width 100%
 - <b>middle</b>: String
    - Loader size height 240px width 100%
 - <b>small</b>: String
    - Loader size height 170px width 100%

### Default
- middle

```html
<Card sizeLoader='small' />
```

<hr/>

## Card with error message:

 ### Expected

- String or Object

```html
<Card isError='true' />
```

<hr/>

## Card with information:

### Expected
 - Array objects (news)

```html
<Card news=[{},{},{},...] />
```

<hr/>

## Card with custom styles:

### Expected

 - <b>styles</b>: Add styles to the main component Card
    -  Object with JSX Styles
 - <b>stylesImage</b>: Object with JSX Styles
    - Add styles to Image
 - <b>stylesTitle</b>: Object with JSX Styles
    - Add styles to title
 - <b>imageWidth</b>: String
    - Change width to the image
 - <b>imageHeight</b>: String
    - Change height to the image


```html
<Card styles={...} stylesImage={...} stylesTitle={...} imageWidth='300px' imageHeight='300px' />
```