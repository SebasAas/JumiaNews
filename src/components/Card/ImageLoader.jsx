import React from "react"
import ContentLoader from "react-content-loader"
import { secondaryWhite, thirthyWhite, thirthyGrey, fourthyDark } from '../../theme/colors'

const ImageLoader = (props) => {
  const { height, width } = props;
  const currentTheme = localStorage.getItem('theme');

  return (
    <ContentLoader
      speed={2}
      width={width}
      height={height}
      viewBox={`0 0 ${width} ${height}`}
      backgroundColor={currentTheme === 'light' ? secondaryWhite : thirthyGrey}
      foregroundColor={currentTheme === 'light' ? thirthyWhite : fourthyDark}
      {...props}
    >
      <rect x="0" y="0" rx="0" ry="0" width={width} height={height} />
    </ContentLoader>
  )
}

export default ImageLoader