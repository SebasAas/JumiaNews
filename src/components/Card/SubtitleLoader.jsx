import React from "react"
import ContentLoader from "react-content-loader"

import { secondaryWhite, thirthyWhite, thirthyGrey, fourthyDark } from '../../theme/colors'

const SubtitleLoader = (props) => {
  const currentTheme = localStorage.getItem('theme');

  return (
    <ContentLoader
      speed={2}
      width='100%'
      height={80}
      viewBox="0 0 350 80"
      // backgroundColor="#f3f3f3"
      // foregroundColor="#ecebeb"
      backgroundColor={currentTheme === 'light' ? secondaryWhite : thirthyGrey}
      foregroundColor={currentTheme === 'light' ? thirthyWhite : fourthyDark}
      {...props}
    >
      <rect x="6" y="12" rx="3" ry="3" width="286" height="13" />
      <rect x="6" y="53" rx="0" ry="0" width="310" height="13" />
      <rect x="6" y="32" rx="0" ry="0" width="299" height="13" />
    </ContentLoader>
  )
}

export default SubtitleLoader