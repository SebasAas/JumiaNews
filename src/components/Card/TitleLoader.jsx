import React from "react"
import ContentLoader from "react-content-loader"

import { secondaryWhite, thirthyWhite, thirthyGrey, fourthyDark } from '../../theme/colors'

const TitleLoader = (props) => {

  const currentTheme = localStorage.getItem('theme');

  return (
    <ContentLoader
      speed={2}
      width='100%'
      height={60}
      viewBox="0 0 315 60"
      // backgroundColor="#f3f3f3"
      // foregroundColor="#ecebeb"
      backgroundColor={currentTheme === 'light' ? secondaryWhite : thirthyGrey}
      foregroundColor={currentTheme === 'light' ? thirthyWhite : fourthyDark}
      {...props}
    >
      <rect x="21" y="-2" rx="3" ry="3" width="238" height="30" />
      <rect x="41" y="33" rx="0" ry="0" width="200" height="28" />
    </ContentLoader>
  )
}

export default TitleLoader