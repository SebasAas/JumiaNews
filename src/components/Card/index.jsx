import React from 'react'

// Styles
import styled from 'styled-components';

// Loaders
import TitleLoader from './TitleLoader'
import SubtitleLoader from './SubtitleLoader'
import ImageLoader from './ImageLoader'

// Routes
import { Link } from "react-router-dom";

function Card({ news, pathname, loading = false, styles, stylesImage, sizeLoader = 'middle', stylesTitle, isError, imageWidth, imageHeight }) {


  const handleSizeLoader = () => {
    if (sizeLoader === 'big') {
      return (
        <ImageLoader data-testid="image-loader" height='370px' width='100%' />
      )
    } else if (sizeLoader === 'small') {
      return (
        <ImageLoader data-testid="image-loader" height='170px' width='100%' />
      )
    } else {
      return (
        <ImageLoader data-testid="image-loader" height='240px' width='100%' />
      )
    }
  }

  const sourceImage = () => {
    // image link to be attached will be different depending on the request
    // if the information comes from 'home' (top stories) then the img has the full url to the image
    // if the information comes from 'section' (article) then the img hasn't the full url and we need to add  https://www.nytimes.com/
    // if the new doesn't have multimedia, then we render a default image of The New York Time
    if (!news?.multimedia) {
      return 'https://upload.wikimedia.org/wikipedia/commons/thumb/4/40/New_York_Times_logo_variation.jpg/1125px-New_York_Times_logo_variation.jpg'
    }
    if (news?.multimedia.length === 0) {
      return 'https://upload.wikimedia.org/wikipedia/commons/thumb/4/40/New_York_Times_logo_variation.jpg/1125px-New_York_Times_logo_variation.jpg'
    }
    if (news?.multimedia[0]?.url.includes('http')) {
      return `${news?.multimedia[0]?.url}`
    } else if (!news?.multimedia[0]?.url.includes('http')) {
      return `https://www.nytimes.com/${news?.multimedia[0]?.url}`
    } else {
      return 'https://upload.wikimedia.org/wikipedia/commons/thumb/4/40/New_York_Times_logo_variation.jpg/1125px-New_York_Times_logo_variation.jpg'
    }
  }

  return (
    <CardContainer className="card__container" style={{ ...styles }}>
      {
        loading ?
          <>
            <TextLoader>
              <TitleLoader data-testid="title-loader" />
              <SubtitleLoader data-testid="subtitle-loader" />
            </TextLoader>
            <ImageLoaderContainer>
              {handleSizeLoader()}
            </ImageLoaderContainer>
          </>
          :
          <>
            <Link
              // Remove all special characters and encode url
              to={`/news/${news?.title ? encodeURI(news?.title?.replace(/[^\w\s]/gi, '')) : encodeURI(news?.headline?.main?.replace(/[^\w\s]/gi, ''))}`}
              state={{ from: pathname, news }}
              style={{ width: '100%', height: 'max-content' }}
            >
              <Text>
                <Title style={{ ...stylesTitle }}>{news?.headline?.main || news?.title}{isError && 'No title found'}</Title>
                <Subtitle>{news?.snippet || news?.abstract}{isError && 'No description found'}</Subtitle>
              </Text>
            </Link>
            <Image
              loading="lazy"
              decoding="async"
              style={{ ...stylesImage }}
              imageWidth={imageWidth}
              imageHeight={imageHeight}
              src={sourceImage()}
            />
          </>
      }
    </CardContainer>
  )
}

const CardContainer = styled.div`
  display: flex;
  flex-direction: column;
  margin: 30px 15px;
  width: 400px;

  @media (max-width: 780px) {
    width: 100% !important;
    height: auto !important;
    flex-direction: column-reverse !important;
  }
`

const Image = styled.img`
    width: ${({ imageWidth }) => imageWidth ? imageWidth : ''};
    height: ${({ imageHeight }) => imageHeight ? imageHeight : ''};
    object-fit: cover;
    min-width: 208px;
    padding: 3px;
    box-shadow: rgba(0, 0, 0, 0.1) 0px 1px 3px 0px, rgba(0, 0, 0, 0.06) 0px 1px 2px 0px;

    @media (max-width: 780px) {
      width: 100%;
      height: auto;
    }
`

const ImageLoaderContainer = styled.div`
  box-shadow: rgb(0 0 0 / 10%) 0px 1px 3px 0px, rgb(0 0 0 / 6%) 0px 1px 2px 0px;
`

const Text = styled.div`
    height: 100%;
    border-top-right-radius: 4px;
    border-top-left-radius: 4px;
    display: flex;
    flex-direction: column;
    align-items: center;
    padding: 0px 20px;
    box-shadow: rgba(0, 0, 0, 0.1) 0px 1px 3px 0px, rgba(0, 0, 0, 0.06) 0px 1px 2px 0px;
`

const TextLoader = styled.div`
    height: 100%;
    min-width: 55%;
    border-top-right-radius: 4px;
    border-top-left-radius: 4px;
    display: flex;
    flex-direction: column;
    align-items: center;
    padding: 25px 40px;
    box-shadow: rgba(0, 0, 0, 0.1) 0px 1px 3px 0px, rgba(0, 0, 0, 0.06) 0px 1px 2px 0px;
`

const Title = styled.h2`
    width: 100%;
    text-align: initial;
    padding-top: 25px;
    display: flex;
    justify-content: center;
    align-items: center;
    min-height: 60px;
    margin: 0 0 10px;
    font-size: 24px;
    line-height: 28px;
    display: -webkit-box;
    -webkit-line-clamp: 2;
    -webkit-box-orient: vertical;
    overflow: hidden;
    text-overflow: ellipsis;

    @media (max-width: 780px) {
      -webkit-line-clamp: 4;
    }
`

const Subtitle = styled.p`
    display: -webkit-box;
    -webkit-line-clamp: 3;
    -webkit-box-orient: vertical;
    overflow: hidden;
    text-overflow: ellipsis;
    font-size: 16px;
    line-height: 20px;
    margin-bottom: 35px;
`

export default React.memo(Card)
