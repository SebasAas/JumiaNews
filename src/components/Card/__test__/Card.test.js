import React from 'react';
import "@testing-library/jest-dom/extend-expect";
import { render } from "@testing-library/react";

import {
  queryByTestId,
} from '@testing-library/dom'

import { MemoryRouter } from 'react-router-dom'

import 'jest-styled-components'

// Components
import Card from '../index'

describe('Component Card is Rendering', () => {

  afterEach(() => {
    jest.resetAllMocks();
  });
  afterAll(() => {
    jest.clearAllMocks();
  });
  beforeEach(() => {
    window.scrollTo = jest.fn();
  });

  test('Should render Default Card if no news are passed', () => {
    const { getByRole } = render(<Card />, { wrapper: MemoryRouter })
    expect(getByRole('img')).toHaveAttribute('src', 'https://upload.wikimedia.org/wikipedia/commons/thumb/4/40/New_York_Times_logo_variation.jpg/1125px-New_York_Times_logo_variation.jpg')
  })

  test('Should render Card Loaders if the loading state is true', () => {
    const { getByTestId } = render(<Card loading={true} />, { wrapper: MemoryRouter });

    // Check if the loaders exist in the doc
    expect(getByTestId('title-loader')).toBeInTheDocument();
    expect(getByTestId('subtitle-loader')).toBeInTheDocument();
    expect(getByTestId('image-loader')).toBeInTheDocument();
  })

  test('Should render Card with information fetched from TopStories (Hightlight)', () => {
    const origin = 'home'
    const news = {
      title: "Title News",
      abstract: "Subtitle News",
      multimedia: [
        {
          url: "https://static01.nyt.com/images/2022/01/15/us/politics/15voting-rights-3/15voting-rights-3-superJumbo.jpg"
        }
      ]
    }

    const { getByText, getByRole } = render(<Card news={news} origin={origin} />, { wrapper: MemoryRouter });

    expect(getByText(news.title)).toBeVisible();
    expect(getByText(news.abstract)).toBeVisible();
    expect(getByRole('img')).toHaveAttribute('src', 'https://static01.nyt.com/images/2022/01/15/us/politics/15voting-rights-3/15voting-rights-3-superJumbo.jpg')
  })

  test('Should render Card with information fetched from Sections', () => {
    const origin = 'section'
    const news = {
      headline: {
        main: "Title News"
      },
      snippet: "Subtitle News",
      multimedia: [
        {
          url: "images/2022/01/16/business/15NFTGAMERS/merlin_200341866_d4bb783c-85f7-4b11-9ce7-7ea77ba7365b-articleLarge.jpg"
        }
      ]
    }

    const { getByText, getByRole } = render(<Card news={news} origin={origin} />, { wrapper: MemoryRouter });

    expect(getByText(news.headline.main)).toBeVisible();
    expect(getByText(news.snippet)).toBeVisible();
    expect(getByRole('img')).toHaveAttribute('src', 'https://www.nytimes.com/images/2022/01/16/business/15NFTGAMERS/merlin_200341866_d4bb783c-85f7-4b11-9ce7-7ea77ba7365b-articleLarge.jpg')
  })

  test('Should render Card Image with custom styles', () => {
    const imageWidth = '120px'
    const imageHeight = '120px'
    const stylesImage = {
      objectFit: 'container'
    }

    const { getByRole } = render(<Card stylesImage={stylesImage} imageWidth={imageWidth} imageHeight={imageHeight} />, { wrapper: MemoryRouter });

    expect(getByRole('img')).toHaveStyle({
      objectFit: stylesImage.objectFit,
      height: imageHeight,
      width: imageWidth
    })
  })

  test('Should render Card Title with custom styles', () => {
    const stylesTitle = {
      color: 'red'
    }

    const { getByText } = render(<Card stylesTitle={stylesTitle} isError={true} />, { wrapper: MemoryRouter });

    expect(getByText('No title found')).toHaveStyle({
      color: stylesTitle.color
    })
  })

  test('Should render Card Title No Title Found on error', () => {
    const { getByText } = render(<Card isError={true} />, { wrapper: MemoryRouter });

    expect(getByText('No title found')).toBeVisible();
  })

  test('Should render Card Description No Description Found on error', () => {
    const { getByText } = render(<Card isError={true} />, { wrapper: MemoryRouter });

    expect(getByText('No description found')).toBeVisible();
  })
  // isError

  test('Should render the Card Loader at first and then re-render with new information', () => {
    const origin = 'home'
    const news = {
      title: "Title News",
      abstract: "Subtitle News",
      multimedia: [
        {
          url: "https://static01.nyt.com/images/2022/01/15/us/politics/15voting-rights-3/15voting-rights-3-superJumbo.jpg"
        }
      ]
    }

    const { rerender, container, getByText, getByRole, getByTestId } = render(<Card loading={true} />, { wrapper: MemoryRouter });
    // Check if the loaders exist in the doc
    expect(getByTestId('title-loader')).toBeInTheDocument();
    expect(getByTestId('subtitle-loader')).toBeInTheDocument();
    expect(getByTestId('image-loader')).toBeInTheDocument();

    // Re render passing new props
    rerender(<Card loading={false} origin={origin} news={news} />, { wrapper: MemoryRouter })

    // Check if the loaders have been remove
    expect(queryByTestId(container, 'title-loader')).toBeNull();
    expect(queryByTestId(container, 'subtitle-loader')).toBeNull();
    expect(queryByTestId(container, 'image-loader')).toBeNull();

    // Check if the news content is visible
    expect(getByText(news.title)).toBeVisible();
    expect(getByText(news.abstract)).toBeVisible();
    expect(getByRole('img')).toHaveAttribute('src', 'https://static01.nyt.com/images/2022/01/15/us/politics/15voting-rights-3/15voting-rights-3-superJumbo.jpg')
  })

  test('should match snapshot card information', () => {
    const origin = 'home'
    const news = {
      title: "Title News",
      abstract: "Subtitle News",
      multimedia: [
        {
          url: "https://static01.nyt.com/images/2022/01/15/us/politics/15voting-rights-3/15voting-rights-3-superJumbo.jpg"
        }
      ]
    }

    const { container } = render(<Card news={news} origin={origin} />, { wrapper: MemoryRouter });

    expect(container).toMatchSnapshot()
  })
})
