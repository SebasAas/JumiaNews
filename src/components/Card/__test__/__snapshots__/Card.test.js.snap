// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Card is Rendering should match snapshot card information 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 30px 15px;
  width: 400px;
}

.c4 {
  object-fit: cover;
  min-width: 208px;
  padding: 3px;
  box-shadow: rgba(0,0,0,0.1) 0px 1px 3px 0px,rgba(0,0,0,0.06) 0px 1px 2px 0px;
}

.c1 {
  height: 100%;
  border-top-right-radius: 4px;
  border-top-left-radius: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0px 20px;
  box-shadow: rgba(0,0,0,0.1) 0px 1px 3px 0px,rgba(0,0,0,0.06) 0px 1px 2px 0px;
}

.c2 {
  width: 100%;
  text-align: initial;
  padding-top: 25px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 60px;
  margin: 0 0 10px;
  font-size: 24px;
  line-height: 28px;
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c3 {
  display: -webkit-box;
  -webkit-line-clamp: 3;
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;
  font-size: 16px;
  line-height: 20px;
  margin-bottom: 35px;
}

@media (max-width:780px) {
  .c0 {
    width: 100% !important;
    height: auto !important;
    -webkit-flex-direction: column-reverse !important;
    -ms-flex-direction: column-reverse !important;
    flex-direction: column-reverse !important;
  }
}

@media (max-width:780px) {
  .c4 {
    width: 100%;
    height: auto;
  }
}

@media (max-width:780px) {
  .c2 {
    -webkit-line-clamp: 4;
  }
}

<div>
  <div
    class="c0 card__container"
  >
    <a
      href="/news/Title%20News"
      style="width: 100%;"
    >
      <div
        class="c1"
      >
        <h2
          class="c2"
        >
          Title News
        </h2>
        <p
          class="c3"
        >
          Subtitle News
        </p>
      </div>
    </a>
    <img
      class="c4"
      decoding="async"
      loading="lazy"
      src="https://static01.nyt.com/images/2022/01/15/us/politics/15voting-rights-3/15voting-rights-3-superJumbo.jpg"
    />
  </div>
</div>
`;
